packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -202,10 +202,15 @@ class DataSourceWithBackend<
     headers[PluginRequestHeaders.PluginID] = Array.from(pluginIDs).join(', ');
     headers[PluginRequestHeaders.DatasourceUID] = Array.from(dsUIDs).join(', ');
 
-    let url = '/api/ds/query';
+    let url = '/api/ds/query?ds_type=' + this.type;
+
     if (hasExpr) {
       headers[PluginRequestHeaders.FromExpression] = 'true';
-      url += '?expression=true';
+      url += '&expression=true';
+    }
+
+    if (requestId) {
+      url += `&requestId=${requestId}`;
     }
```

Comment:
as far as I can tell, data sources cannot access URL params set in the query and that's why "from expression" was added both as a header and a URL param in https://github.com/grafana/grafana/pull/65306. Why not using headers here to avoid ending in the same situation?

Reply:
I believe this is because in the `PerformanceObserver` all we have access to is the url, and we don't have access to headers in that context. @leeoniya is that correct?

Reply:
yes, this is purely for frontend tracking. these requestIds are not unique, they simply increment from initial dashboard load (Q100, Q101, Q102...). when we construct these requests we need to somehow correlate them from an async PerformanceObserver callback so we can measure stuff (like transfer savings). unfortunately PerformanceObserver entries don't give us access to anything except the query URL, so this is the only place we can stick it to make that link. you can see what we do with it in https://github.com/grafana/grafana/pull/65244

these ids aren't useful in the backend, but if they ever become UIDs and useful, we can always add the header as well in the future.

Reply:
@ryantxu ^^ This is the direct opposite of your request to make this available for all data sources.

Reply:
@gtk-grafana already has an Influx PR that will use this also, next will be loki/logs. after that it can be any datasource that returns timestamps in the data.

in general, this strategy allows us to cheaply collect perf metrics of any DS request without delegating this to the backend which in most cases is infeasible since they stream responses and cannot measure end-to-end performance as the user experiences it.

Reply:
I don't mean to block this, if it will be used widely please go ahead. Also, please evaluate the suggestion at https://github.com/grafana/grafana/pull/65244#issuecomment-1496161971 in case it's helpful.